wasmtime-wasi/src/host_impls.rs

```diff
@@ -1,5 +1,65 @@
+use super::host;
 use super::wasm32;
 
+pub unsafe fn wasmtime_ssp_args_get(
+    argv_environ: &mut host::argv_environ_values,
+    argv: &mut [*mut host::char],
+    argv_buf: *mut host::char,
+) -> wasm32::__wasi_errno_t {
+    for i in 0..(*argv_environ).argc {
+        let buf_off = isize::checked_sub(
+            *(*argv_environ).argv.offset(i as isize) as _,
+            (*argv_environ).argv_buf as _,
+        )
+        .expect("argv[i] - argv_buf overflows");
+        argv[i] = argv_buf.offset(buf_off);
+    }
+    argv[(*argv_environ).argc] = std::ptr::null_mut();
+    argv_buf.copy_from((*argv_environ).argv_buf, (*argv_environ).argv_buf_size);
+    wasm32::__WASI_ESUCCESS
```

Comment:
Technically, the return value from any of the host syscalls should be `host::__wasi_errno_t`. This would agree with the fact that `syscalls::return_encoded_errno` takes `host::__wasi_errno_t` as input arg and return `wasm32::__wasi_errno_t`. I'm wondering though whether the host syscalls we rewrite in Rust should follow that as well. It looks like we'd repeat ourselves a lot by essentially copying over the majority of the contents of `wasm32`. I've posited that question here as in other syscalls handling errors becomes significant and I'm wondering what the best strategy to handle it would be. What do you guys think?

Reply:
posix.c is peppered with `convert_errno` calls, so the `host_impls` version would be doing errno conversion too; it seems a little convoluted to translate errno values to `host::__wasi_errno_t` in `host_impls` then convert to `wasm32::__wasi_errno_t` later on instead of using `wasm32` from the start.

Reply:
Yep, good point. On closer inspection, I see that the `host::__wasi_errno_t` and `wasm32::__wasi_errno_t` are identical except for the size of the integer. However, `convert_errno` from POSIX to WASI will have to be reimplemented as, AFAIK, there isn't a direct conversion between the two; e.g., `EIO=5` in POSIX while `EIO=29` in WASI.

Personally, I'm inclined to have the host syscalls return `wasm32::__wasi_errno_t` directly, as per your suggestion. What are your thoughts @sunfishcode on this?

Reply:
Having the host functions operate in terms of the host types exclusively has some nice properties. It's more consistent, helps anticipate adding wasm64, makes it easier to work on host_impls.rs without remembering which things are host::* and which are wasm32::*, and makes it possible to write tests that don't use any wasm and just call host functions natively.

Doing errno translation in two phases -- errno -> __wasi_errno_t and then host to wasm32, isn't that much of a burden. There's only one call to `encode_errno` in syscalls.rs, in `return_encoded_errno`, which also logs the errno value.

So unless someone feels strongly, I'd prefer to keep host_impls.rs independent of wasm32 :-).

Reply:
SGTM! Have a look at #127 then, and see if that's closer to what we're after :-)